File: advance/src/eu/advance/logistics/flow/engine/api/ds/AdvanceJMSEndpoint.java
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */

package eu.advance.logistics.flow.engine.api.ds;

import hu.akarnokd.reactive4java.base.Func0;
import eu.advance.logistics.flow.engine.api.core.Copyable;
import eu.advance.logistics.flow.engine.api.core.HasPassword;
import eu.advance.logistics.flow.engine.api.core.Identifiable;
import eu.advance.logistics.flow.engine.xml.typesystem.XElement;
import eu.advance.logistics.flow.engine.xml.typesystem.XSerializable;

/**
 * Definition of a Java Messaging Service endpoint.
 * @author akarnokd, 2011.09.20.
 */
public class AdvanceJMSEndpoint extends AdvanceCreateModifyInfo 
implements XSerializable, HasPassword, Copyable<AdvanceJMSEndpoint>, Identifiable<String> {
	/** The name used by blocks to reference this endpoint. */
	public String name;
	/** The JMS driver. */
	public String driver;
	/** The connection URL. */
	public String url;
	/** The user name. */
	public String user;
	/**
	 * The password. 
	 * <p>Note that passwords are never returned from the 
	 * control API calls and are always {@code null}.</p> 
	 * <p>An empty password should be an empty {@code char} array. To keep
	 * the current password, use {@code null}.</p>
	 */
	private char[] password;
	/** The queue manager name. */
	public String queueManager;
	/** The queue name. */
	public String queue;
	/** The communication pool size. */
	public int poolSize;
	/** The function to create a new instance of this class. */
	public static final Func0<AdvanceJMSEndpoint> CREATOR = new Func0<AdvanceJMSEndpoint>() {
		@Override
		public AdvanceJMSEndpoint invoke() {
			return new AdvanceJMSEndpoint();
		}
	};
	@Override
	public void load(XElement source) {
		name = source.get("name");
		driver = source.get("driver");
		url = source.get("url");
		user = source.get("user");
		password = getPassword(source, "password");
		
		queueManager = source.get("queue-manager");
		queue = source.get("queue");
		poolSize = source.getInt("poolsize");
		
		super.load(source);
	}
	@Override
	public void save(XElement destination) {
		destination.set("name", name);
		destination.set("driver", driver);
		destination.set("url", url);
		destination.set("user", user);
		setPassword(destination, "password", password);
		destination.set("queue-manager", queueManager);
		destination.set("queue", queue);
		destination.set("poolsize", poolSize);
		
		super.save(destination);
	}
	@Override
	public AdvanceJMSEndpoint copy() {
		AdvanceJMSEndpoint result = new AdvanceJMSEndpoint();
		
		result.name = name;
		result.driver = driver;
		result.url = url;
		result.user = user;
		result.queueManager = queueManager;
		result.queue = queue;
		result.poolSize = poolSize;
		result.password = password != null ? password.clone() : null;
		
		assignTo(result);
		
		return result;
	}
	@Override
	public char[] password() {
		return password != null ? password.clone() : null;
	}
	@Override
	public void password(char[] newPassword) {
		password = newPassword != null ? newPassword.clone() : null;
	}
	@Override
	public String id() {
		return name;
	}
}

File: advance-api-cs/AdvanceAPIClient/Classes/AdvanceJMSEndpoint.cs
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Globalization;
using System.Xml;

using AdvanceAPIClient.Core;
using AdvanceAPIClient.Interfaces;

namespace AdvanceAPIClient.Classes
{
    public class AdvanceJMSEndpoint : AdvanceCreateModifyInfo, IPassword, ICopiable<AdvanceJMSEndpoint>, IIdentifiable<string>
{
        /// <summary>
        /// name used by blocks to reference this data source
        /// </summary>
        public string Id { get { return this.Name; } }
        public string Name;
        /// <summary>
        /// JDBC driver
        /// </summary>
        public string Driver;
        /// <summary>
        /// connection url
        /// </summary>
        public String Url;
        /// <summary>
        /// User name used for login
        /// </summary>
        public string User;
        /// <summary>
        /// User's password.
        /// <remarks>
        /// Note that passwords are never returned from the 
        /// control API calls and are always null
        /// An empty password should be an char[0] array. To keep
        /// the current password, use null
        /// </remarks>
        /// </summary>
        public char[] Password
        {
            get { return (this.password == null) ? null : (char[])this.password.Clone(); }
            set { this.password = (value == null) ? null : (char[])value.Clone(); }
        }
        private char[] password;
        /// <summary>
        /// Queue manager name
        /// </summary>
        public string QueueManager;
        /// <summary>
        /// Queue name
        /// </summary>
          public string Queue;
        /// <summary>
        /// communication pool size
        /// </summary>
        public int PoolSize;

        public AdvanceJMSEndpoint() : base() { }

        /// <summary>
        /// Fill object data from xml node
        /// </summary>
        /// <param name="source">Source Xml</param>
        protected override void LoadFromXmlNode(XmlNode source)
        {
            base.LoadFromXmlNode(source);
            this.Name = GetAttribute(source, "name");
            this.User = GetAttribute(source, "user", null);
            this.password = GetPassword(source, "password");
            this.Driver = GetAttribute(source, "driver", null);
            this.Url = GetAttribute(source, "url", null);
            this.PoolSize = GetIntAttribute(source, "poolsize", 5);
            this.Queue = GetAttribute(source, "queue", null);
            this.QueueManager = GetAttribute(source, "queue_manager", null);
        }

        /// <summary>
        /// Fills  object's xml
        /// </summary>
        /// <param name="node">Empty node for the object with name "tagname"</param>
        protected override void FillXmlElement(XmlElement node)
        {
            base.FillXmlElement(node);
            AddAttribute(node, "name", this.Name);
            AddAttribute(node, "user", this.User);
            SetPassword(node, "password", password);
            AddAttribute(node, "driver", this.Driver);
            AddAttribute(node, "url", this.Url);
            AddAttribute(node, "poolsize", this.PoolSize);
            AddAttribute(node, "queue", this.Queue);
            AddAttribute(node, "queue_manager", this.QueueManager);
        }

        public AdvanceJMSEndpoint Copy()
        {
            AdvanceJMSEndpoint result = new AdvanceJMSEndpoint();
            this.AssignTo(result);
            result.Name = this.Name;
            result.password = this.Password;
            result.Driver = this.Driver;
            result.Url = this.Url;
            result.PoolSize = this.PoolSize;
            result.Queue = this.Queue;
            result.QueueManager = this.QueueManager;
            return result;
        }

    }
}
